pandas/src/join.pyx

```diff
@@ -125,6 +125,153 @@ def left_outer_join(ndarray[int64_t] left, ndarray[int64_t] right,
 
 
 
+def left_outer_asof_join(ndarray[int64_t] left, ndarray[int64_t] right,
+                         Py_ssize_t max_groups, sort=True,
+                         bint allow_exact_matches=1,
+                         left_distance=None,
+                         right_distance=None,
+                         tolerance=None):
+
+    cdef:
+        Py_ssize_t i, j, k, count = 0
+        Py_ssize_t loc, left_pos, right_pos, position
+        Py_ssize_t offset
+        ndarray[int64_t] left_count, right_count
+        ndarray left_sorter, right_sorter, rev
+        ndarray[int64_t] left_indexer, right_indexer
+        int64_t lc, rc, tol, left_val, right_val, diff, indexer
+        ndarray[int64_t] ld, rd
+        bint has_tol = 0
+
+    # if we are using tolerance, set our objects
+    if left_distance is not None and right_distance is not None and tolerance is not None:
+        has_tol = 1
+        ld = left_distance
+        rd = right_distance
+        tol = tolerance
+
+    # NA group in location 0
+    left_sorter, left_count = groupsort_indexer(left, max_groups)
+    right_sorter, right_count = groupsort_indexer(right, max_groups)
```

Comment:
Have you looked into performance optimizations for "well-behaved" time series (monotonic, no NAs)? In such cases you may be able to avoid both the groupsort and factorize (both of which are O(n), but can introduce significant / unnecessary overhead)